clang/lib/Sema/CheckExprLifetime.cpp

```diff
@@ -1110,13 +1117,14 @@ static bool shouldRunGSLAssignmentAnalysis(const Sema &SemaRef,
            isAssignmentOperatorLifetimeBound(Entity.AssignmentOperator)));
 }
 
-static void checkExprLifetimeImpl(Sema &SemaRef,
-                                  const InitializedEntity *InitEntity,
-                                  const InitializedEntity *ExtendingEntity,
-                                  LifetimeKind LK,
-                                  const AssignedEntity *AEntity, Expr *Init) {
-  assert((AEntity && LK == LK_Assignment) ||
-         (InitEntity && LK != LK_Assignment));
+static void
+checkExprLifetimeImpl(Sema &SemaRef, const InitializedEntity *InitEntity,
+                      const InitializedEntity *ExtendingEntity, LifetimeKind LK,
+                      const AssignedEntity *AEntity,
+                      const CapturingEntity *CapEntity, Expr *Init) {
+  assert(!AEntity || LK == LK_Assignment);
+  assert(!CapEntity || LK == LK_LifetimeCapture);
+  assert(!InitEntity || (LK != LK_Assignment && LK != LK_LifetimeCapture));
```

Comment:
Yeah. That is a valid concern. I had previously proposed to change this to a class and have separate member functions instead of passing these lambdas in each function call (specially the temporary visitor).

This was added as a FIXME just few lines below:
```
  // FIXME: consider moving the TemporaryVisitor and visitLocalsRetained*
  // functions to a dedicated class.
```
This is a larger refactoring. I prefer to do this in a separate PR.

Reply:
I agree that `checkExprLifetimeImpl` is quite complex, especially with the 300-line lambda. However, this change doesn't significantly worsen the situation, as it only adds a few lines of code to handle the new case. Considering the minimal impact, I’d prefer not to block this patch.

That said, I understand the importance of reducing overall complexity. I think we can  refactor it in a follow-up.

Reply:
I can commit to do this in a follow-up.